crates/forge-runner/src/running.rs

```diff
@@ -277,15 +308,60 @@ pub fn run_test_case(
         None
     };
 
+    // copied from casm_run
     let mut vm = VirtualMachine::new(true);
-    let run_result = runner.run_function_with_vm(
-        func,
+    let data: Vec<MaybeRelocatable> = instructions
+        .flat_map(|inst| inst.assemble().encode())
+        .map(Felt252::from)
+        .map(MaybeRelocatable::from)
+        .collect();
+    let data_len = data.len();
+    // end of copied code
+
+    let mut runner = casm_run::build_runner(data, builtins, hints_dict)?;
+    let run_result = match casm_run::run_function_with_runner(
         &mut vm,
+        data_len,
+        initialize_vm,
         &mut forge_runtime,
-        hints_dict,
-        instructions,
-        builtins,
-    );
+        &mut runner,
+    ) {
+        Ok(()) => {
+            finalize(
+                &mut vm,
+                &runner,
+                &mut forge_runtime
+                    .extended_runtime
+                    .extended_runtime
+                    .extended_runtime
+                    .hint_handler,
+                0,
+                2,
+            );
```

Comment:
I'm not sure if we need `finalize` anymore. I think we can simply call `get_execution_resources` on the `runner` now

Reply:
Tried to use `finalize_execution` from blockifier directly instead, but it takes all arguments by value instead of reference and we still need access to `forge_runtime` after finalization.

Reply:
@cptartur can't we do it last then?

Reply:
We'd have to clone `cells` we get from the casm runner even if we did it last. Because otherwise the compiler complains about it being partially moved.

I'll try thinking of the way of using `finalize_execution` but it's not trivial.

Reply:
Checked again, and the biggest problem is that `finalize_execution` needs to take `SyscallHintProcessor` by value, and we still need access to `forge_runtime` after finalization, so we can extract execution resources from it.

Reply:
Removed some code from finalize and inlined the rest of it (adding resources). It should be simpler now.